File: validate-binary-search-tree/validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
    return isValid(root, nil, nil)
}

func isValid(node *TreeNode, min *int, max *int) bool{
    if node == nil{
        return true
    }
    value := node.Val
    
    if min != nil{
        if value <= *min{
            return false
        }
    }
    if max != nil{
        if value >= *max{
            return false
        }
    }
    return isValid(node.Left, min, &value) && isValid(node.Right, &value, max)
}

File: validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        
        def helper(node, lower, upper):
            
            if not node:
				# empty node or empty tree
                return True
            
            if lower < node.val < upper:
				# check if all tree nodes follow BST rule
                return helper(node.left, lower, node.val) and helper(node.right, node.val, upper)
            
            else:
				# early reject when we find violation
                return False
            
        # ----------------------------------
        
        return helper( node=root, lower=float('-inf'), upper=float('inf') )